funsor/tensor.py

```diff
@@ -484,10 +506,26 @@ def align_tensors(*args, **kwargs):
 
 
 @to_data.register(Tensor)
-def _to_data_tensor(x):
-    if x.inputs:
-        raise ValueError(f"cannot convert Tensor to data due to lazy inputs: {set(x.inputs)}")
-    return x.data
+def _to_data_tensor(x, name_to_dim=None):
+    if not name_to_dim or not x.inputs:
+        if x.inputs:
+            raise ValueError(f"cannot convert Tensor to data due to lazy inputs: {set(x.inputs)}")
+        return x.data
+    else:
```

Comment:
nit: could you add an assertion about `name_to_dim` similar to the assertion above in `tensor_to_funsor()`?